### DISCUSSION TITLE
Refactor non-bonded interaction types

### DISCUSSION THREAD

**Discussion by KaiSzuttor** [espressomd/espresso#4214] [Ideas]
The parameters and force and energy calculations of the non-bonded interactions are currently split
and could be combined in a new type. E.g. the LJ interaction could have the following layout:

``` c++
template <class T> struct LennardJones {
   using value_type = T;
   value_type epsilon;
   value_type sigma;

   value_type energy(value_type r) const {
     return 4.0 * epsilon *
            (Utils::int_pow<12>(sigma / r) - Utils::int_pow<6>(sigma / r));
   }

   Utils::Vector<value_type, 3>
   force(value_type r, Utils::Vector<value_type, 3> const &r12) const {
     return r12 *
            (24. * epsilon * Utils::int_pow<6>(sigma) *
             (Utils::int_pow<6>(r) - 2. * Utils::int_pow<6>(sigma))) /
            Utils::int_pow<13>(r);
   }
 };
```

**Comment by jngrad** [comment]
I think it's a good idea and would conclude the non-bonded part of the refactoring plan proposed in #3037. However, I do not immediately see the benefit of using a templated struct, since we currently only need doubles in the core, and the Cython interface makes it tricky to work with templated types.
